File: LeetCode/cpp/438.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        int n = s.size(), m = p.size();
        if (n < m)
            return {};
        vector<int> std(26, 0), cur(26, 0), ret;
        for (auto ch:p)
            ++std[ch - 'a'];
        for (int i = 0; i < m - 1; ++i)
            ++cur[s[i] - 'a'];
        for (int i = m - 1; i < n; ++i) {
            ++cur[s[i] - 'a'];
            if (cur == std)
                ret.push_back(i - m + 1);
            --cur[s[i - m + 1] - 'a'];
        }
        return ret;
    }
};

File: LeetCode/python3/438.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        m, n = len(p), len(s)
        if m > n:
            return []
        dict, supp = {}, {}
        ret = []
        for i in range(m):
            supp[p[i]] = 1 if p[i] not in supp else supp[p[i]] + 1
        for i in range(m - 1):
            dict[s[i]] = 1 if s[i] not in dict else dict[s[i]] + 1
        for i in range(m - 1, n):
            dict[s[i]] = 1 if s[i] not in dict else dict[s[i]] + 1
            if dict == supp:
                ret.append(i - m + 1)
            dict[s[i - m + 1]] -= 1
            if dict[s[i - m + 1]] == 0:
                del dict[s[i - m + 1]]
        return ret